app/src/main/java/io/stormbird/wallet/service/KeyService.java

```diff
@@ -0,0 +1,1412 @@
+                        default:
+                            if (encryptedHDFileName.contains(NO_AUTH_LABEL)) hdKey.authLevel = AuthenticationLevel.TEE_NO_AUTHENTICATION;
+                            else hdKey.authLevel = AuthenticationLevel.TEE_AUTHENTICATION;
+                            break;
+                        case HAS_STRONGBOX:
+                            if (encryptedHDFileName.contains(NO_AUTH_LABEL)) hdKey.authLevel = AuthenticationLevel.STRONGBOX_NO_AUTHENTICATION;
+                            else hdKey.authLevel = AuthenticationLevel.STRONGBOX_AUTHENTICATION;
+                            break;
+                    }
+
+                    walletMap.put(date, hdKey);
+                }
+            }
+
+            Collections.sort(fileDates);
+
+            for (Date d : fileDates)
+            {
+                wallets.add(walletMap.get(d));
+            }
+        }
+        catch (Exception e)
+        {
+            e.printStackTrace();
+        }
+
+        return wallets;
+    }
+
+    private String createEncryptedFilePath(String address, Operation operation)
+    {
+        String fileName = address;
+        switch (operation)
+        {
+            case UPGRADE_HD_KEY:
+            case IMPORT_HD_KEY:
+            case CREATE_HD_KEY:
+                fileName += HDKEY_LABEL;
+                break;
+
+            case RESTORE_NON_AUTHENTICATED_HD_KEY:
+            case CREATE_NON_AUTHENTICATED_KEY:
+                fileName = fileName + NO_AUTH_LABEL + HDKEY_LABEL;
+                break;
+            case CREATE_KEYSTORE_KEY:
+            case CREATE_PRIVATE_KEY:
+                if (deviceIsLocked())
+                {
+                    fileName += KEYSTORE_LABEL;
+                }
+                else
+                {
+                    fileName = fileName + NO_AUTH_LABEL + KEYSTORE_LABEL;
+                }
+                break;
+            case RESTORE_NON_AUTHENTICATED_KS_KEY:
+                fileName = fileName + NO_AUTH_LABEL + KEYSTORE_LABEL;
+                break;
+            case UPGRADE_KEYSTORE_KEY:
+                fileName += KEYSTORE_LABEL;
+                break;
+            default:
+                break;
+        }
+
+        return getFilePath(context, fileName);
+    }
+
+    private boolean keyIsAuthLocked(Operation operation)
+    {
+        switch (operation)
+        {
+            case UPGRADE_HD_KEY:
+            case CREATE_HD_KEY:
+            case UPGRADE_KEYSTORE_KEY:
+            case IMPORT_HD_KEY:
+                return true;
+            case CREATE_NON_AUTHENTICATED_KEY:
+            case RESTORE_NON_AUTHENTICATED_HD_KEY:
+            case RESTORE_NON_AUTHENTICATED_KS_KEY:
+                return false;
+            case CREATE_KEYSTORE_KEY:
+            case CREATE_PRIVATE_KEY:
+                return deviceIsLocked();
+            default:
+                return true;
+        }
+    }
+
+    private String getEncryptedFilePath(String keyAddr)
```

Comment:
It would be more robust to explicitly specify what type of file (i.e., auth-locked, non-auth-locked, etc) we are trying to fetch, and fail if it isn't there. Otherwise, we can run into some bad flows. For example, suppose the attacker has access to the file system. He could then remove an existing HD-auth-locked seed file, and copy back on old HD-non-auth-locked seed file; this method would then silently return the HD-non-auth-locked seed file to the caller, even though the caller expects an HD-auth-locked seed file. 

An even more robust approach would be to tie the key path names directly to the secure enclave, i.e., if the AES key in the enclave is not auth-locked, use "...-no-auth-hd", else use "...-hd".